src/restic/backend/b2/b2.go

```diff
@@ -0,0 +1,288 @@
+	defer r.Close()
+
+	bufsize := int64(len(p))
+	var pos int64 = 0
+
+	// loop to read chunks until the buffer is full or EOF is reached
+	for {
+		n, err = r.Read(p[pos:])
+		pos += int64(n)
+		if pos == info.Size-off || pos == bufsize || err != nil {
+			if errors.Cause(err) == io.EOF {
+				err = nil
+			}
+			break
+		}
+	}
+
+	if err == nil {
+		err = nextError
+	}
+
+	return int(pos), err
+}
+
+// Save stores data in the backend at the handle.
+func (be b2) Save(h restic.Handle, p []byte) (err error) {
+	debug.Log("Save: %v with len %d", h, len(p))
+	if err := h.Valid(); err != nil {
+		return err
+	}
+
+	objName := be.b2path(h.Type, h.Name)
+
+	obj := be.bucket.Object(objName)
+	if err != nil {
+		debug.Log("  err %v", err)
+		return errors.Wrap(err, "bucket.Object")
+	}
+
+	_, err = obj.Attrs(be.context)
+	if err == nil {
+		debug.Log("  %v already exists", h)
+		return errors.New("key already exists")
+	}
+
+	debug.Log("  Writing to %s", objName)
+	w := obj.NewWriter(be.context)
+	n, err := w.Write(p)
+
+	if err != nil {
+		debug.Log("  Write error %v -> %v bytes, err %#v", objName, n, err)
+		w.Close()
+		return err
+	}
+
+	debug.Log("  Write successful %v -> %v bytes", objName, n)
+	w.Close()
+	return nil
+}
+
+// Stat returns information about a blob.
+func (be b2) Stat(h restic.Handle) (bi restic.FileInfo, err error) {
+	debug.Log("Stat: %v", h)
+	objName := be.b2path(h.Type, h.Name)
+	obj := be.bucket.Object(objName)
+	info, err := obj.Attrs(be.context)
+	if err != nil {
+		debug.Log("Attrs() err %v", err)
+		return restic.FileInfo{}, errors.Wrap(err, "Stat")
+	}
+	return restic.FileInfo{Size: info.Size}, nil
+}
+
+// Test returns true if a blob of the given type and name exists in the backend.
+func (be *b2) Test(t restic.FileType, name string) (bool, error) {
+	found := false
+	objName := be.b2path(t, name)
+	obj := be.bucket.Object(objName)
+	_, err := obj.Attrs(be.context)
+	if err == nil {
+		found = true
+	}
+	return found, nil
+}
+
+// Remove removes the blob with the given name and type.
+func (be *b2) Remove(t restic.FileType, name string) error {
+	objName := be.b2path(t, name)
+
+	obj := be.bucket.Object(objName)
+
+	err := obj.Delete(be.context)
+	debug.Log("%v %v -> err %v", t, name, err)
+	return errors.Wrap(err, "object.Delete")
+}
+
+// List returns a channel that yields all names of blobs of type t. A
+// goroutine is started for this. If the channel done is closed, sending
+// stops.
+func (be *b2) List(t restic.FileType, done <-chan struct{}) <-chan string {
+	debug.Log("List: %v", t)
+	ch := make(chan string)
+
+	go func() {
+		defer close(ch)
+		prefix := path.Join(be.prefix, string(t)) + "/"
+		cur := &blazerb2.Cursor{Name: prefix}
+
+		for {
+			objs, c, err := be.bucket.ListCurrentObjects(be.context, 1000, cur)
+			if err != nil && err != io.EOF {
+				return
+			}
+			for _, obj := range objs {
+				info, err := obj.Attrs(be.context)
+				if err != nil {
```

Comment:
Instead of making this change, I've removed the call to `obj.Attrs()` in `List()`, which is no longer necessary for getting the object's name due to your change to blazer in: https://github.com/kurin/blazer/pull/3

This change benefits performance and API usage costs by not making an attributes request for every object returned in a list operation. Additionally, I think it's appropriate for the list method to return objects which exist but are still being uploaded.